source/remoteClient/client.py

```diff
@@ -0,0 +1,483 @@
+			connection_info=connectionInfo,
+			serializer=serializer.JSONSerializer(),
+		)
+		self.masterSession = MasterSession(
+			transport=transport,
+			localMachine=self.localMachine,
+		)
+		transport.transportCertificateAuthenticationFailed.register(
+			self.onMasterCertificateFailed,
+		)
+		transport.transportConnected.register(self.onConnectedAsMaster)
+		transport.transportConnectionFailed.register(self.onConnectAsMasterFailed)
+		transport.transportClosing.register(self.onDisconnectingAsMaster)
+		transport.transportDisconnected.register(self.onDisconnectedAsMaster)
+		transport.reconnectorThread.start()
+		self.masterTransport = transport
+		if self.menu:
+			self.menu.handleConnecting(connectionInfo.mode)
+
+	@alwaysCallAfter
+	def onConnectedAsMaster(self):
+		log.info("Successfully connected as master")
+		configuration.write_connection_to_config(self.masterSession.getConnectionInfo())
+		if self.menu:
+			self.menu.handleConnected(ConnectionMode.MASTER, True)
+		ui.message(
+			# Translators: Presented when connected to the remote computer.
+			_("Connected!"),
+		)
+		cues.connected()
+
+	@alwaysCallAfter
+	def onDisconnectingAsMaster(self):
+		log.info("Master session disconnecting")
+		if self.menu:
+			self.menu.handleConnected(ConnectionMode.MASTER, False)
+		if self.localMachine:
+			self.localMachine.isMuted = False
+		self.sendingKeys = False
+		self.keyModifiers = set()
+
+	@alwaysCallAfter
+	def onDisconnectedAsMaster(self):
+		log.info("Master session disconnected")
+		# Translators: Presented when connection to a remote computer was interupted.
+		ui.message(_("Connection interrupted"))
+
+	def connectAsSlave(self, connectionInfo: ConnectionInfo):
+		transport = RelayTransport.create(
+			connection_info=connectionInfo,
+			serializer=serializer.JSONSerializer(),
+		)
+		self.slaveSession = SlaveSession(
+			transport=transport,
+			localMachine=self.localMachine,
+		)
+		self.sdHandler.slaveSession = self.slaveSession
+		self.slaveTransport = transport
+		transport.transportCertificateAuthenticationFailed.register(
+			self.onSlaveCertificateFailed,
+		)
+		transport.transportConnected.register(self.onConnectedAsSlave)
+		transport.transportDisconnected.register(self.onDisconnectedAsSlave)
+		transport.reconnectorThread.start()
+		if self.menu:
+			self.menu.handleConnecting(connectionInfo.mode)
+
+	@alwaysCallAfter
+	def onConnectedAsSlave(self):
+		log.info("Control connector connected")
+		cues.control_server_connected()
+		if self.menu:
+			self.menu.handleConnected(ConnectionMode.SLAVE, True)
+		configuration.write_connection_to_config(self.slaveSession.getConnectionInfo())
+
+	@alwaysCallAfter
+	def onDisconnectedAsSlave(self):
+		log.info("Control connector disconnected")
+		# cues.control_server_disconnected()
+		if self.menu:
+			self.menu.handleConnected(ConnectionMode.SLAVE, False)
+
+	### certificate handling
+
+	def handleCertificateFailure(self, transport: RelayTransport):
```

Comment:
Does this require an `alwaysCallAfter`?
```suggestion
	@alwaysCallAfter
	def handleCertificateFailure(self, transport: RelayTransport):
```

Reply:
It does not, because its return value is important. Anything it calls does though.